// 接入层C/S协议

package connector_client;

enum ConnectorMagic {
  MAGIC_CS_V1 = 0x3344;
};

enum BodyCmdId {
  SYN              = 1;
  SYNACK           = 2;
  AUTHENTICATION   = 3;
  PASSPORT         = 4;
  DATA_TRANSPARENT = 16;
  ERROR            = 100;
}

message CSHead {
  required int32 magic = 1[default=0x3344];
  required int64 sequence = 2;
  required int32 bodyid = 3;
  required int32 checksum = 4;
}

// 不同时间、不同uin/openid协商出的秘钥不一样
// 如果是自营账号系统，直接采用密码的MD5或二次MD5计算秘钥，无需秘钥协商过程
message CSSyn {
  required string appid  = 1;
  required string openid = 2;
}

message CSSynAck {
  required string secretkey = 1;
}

// 从OAuth2.0的方式来理解的话，token等价于refreshtoken 或 code，
// passport等价于accesstoken
message CSAuth {
  required string token = 1;
}

message CSPassport {
  optional int64 passport = 1;
}

message CSError {
  required int32 code = 1;
  required string desc = 2;
}

// connector做透传，游戏客户端数据作为unknown fields
// 也可以直接透传，bytes类型
message CSBody {
  optional CSSyn syn            = 1;
  optional CSSynAck ack         = 2;
  optional CSAuth auth          = 3;
  optional CSPassport passport  = 4;
  optional bytes data           = 16;
  optional CSError error        = 100;
}

message CSMsg {
  required CSHead head = 1;
  required CSBody body = 2;
}
